### PR TITLE
Support binary literals with binary format

### PR DESCRIPTION
**Pull request by merkrafter**
# Description

This PR lets users specify binary literals in binary format, e.g. `0b[1010 0001]` (before it was only implemented for hex format).
Also, automatic left padding with zeros is enabled, that is `0b[1]` is equivalent to `0b[0000 0001]`.

Relevant documentation update: https://github.com/nushell/nushell.github.io/pull/390

# Tests

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --all --all-features -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style ~-> fails even on main branch at nu-protocol>config.rs:302,339 (config parameter is only used in recursion; clippy 0.1.61 (8f36334 2022-04-06); no complains in my code though)~ works with clippy 0.1.60
- [x] `cargo build; cargo test --all --all-features` to check that all the tests pass

### CODE REVIEW

**Comment by merkrafter** [comment]
Should I also fix the clippy warning even though it's not in scope of my change?

**Comment by sholderbach** [comment]
Which one would that be? Have you rebased onto the latest `main`? The latest PR's seem to pass clippy.

**Comment by merkrafter** [comment]
It's nu-protocol>config.rs:302,339 ('config' parameter is only used in recursion)
I used clippy 0.1.61 (8f36334 2022-04-06); not sure what version runs in your CI at the moment

**Comment by merkrafter** [comment]
Clippy 0.1.60 passed without any problems. Updated PR description.

**Comment by merkrafter** [comment]
Is there any documentation, example, CHANGELOG etc. I need to update?

**Comment by sholderbach** [comment]
The CI runs just `stable`. I  think the clippy fixes should go into a separate PR. Was able to confirm three new warnings with the current `beta` and  two shared and one different warning with a `1.62.0-nightly (e2e2ddd 2022-04-05)`

**Comment by sholderbach** [comment]
For builtin help/examples I am not sure. The general documentation for the web page lives in https://github.com/nushell/nushell.github.io/blob/main/book/types_of_data.md

File: crates/nu-parser/src/parser.rs [hunk 1]
```diff
@@ -1105,20 +1119,21 @@ pub fn parse_binary(
...
+                    binary_value.append(&mut tail);
+                    binary_value
+                };
             }
 
-            let str = String::from_utf8_lossy(&hex_value).to_string();
+            let str = String::from_utf8_lossy(&binary_value).to_string();
 
-            match decode_hex(&str) {
+            match decode_with_base(&str, base, min_digits_per_byte) {
```
Comment (sholderbach) [inline crates/nu-parser/src/parser.rs:1136]: Interesting and sad that the error message thrown by that get's eaten with the eager assumption that everything unknown could be an external command. (I know that it is a valid filename)

File: crates/nu-parser/src/parser.rs [hunk 2]
```diff
@@ -1074,26 +1074,40 @@ pub fn parse_binary(
...
+        return parse_binary_with_base(working_set, span, 2, 8, b"0b[", b"]");
     }
+    (hex_value, err)
+}
 
+fn parse_binary_with_base(
+    working_set: &mut StateWorkingSet,
+    span: Span,
+    base: u32,
+    min_digits_per_byte: usize,
```
Comment (sholderbach) [inline crates/nu-parser/src/parser.rs:1088]: Let's hope they get optimized out in release mode. The `u8::from_str_radix` signature is kind of weird with a base exceeding the valid result type and a smaller range of safe and valid characters.

**Review by sholderbach — APPROVED** [review]
Looks good to me

File: crates/nu-parser/src/parser.rs (hunk 1 above)
Comment (jntrnr) [reply crates/nu-parser/src/parser.rs:1136]: @sholderbach - for which case? I don't quite follow

Comment (sholderbach) [reply crates/nu-parser/src/parser.rs:1136]: Try to put invalid content in the literal body. E.g. `0b[ff]` I got a unknown executable error

Comment (jntrnr) [reply crates/nu-parser/src/parser.rs:1136]: Oooooh, hmm, let me think about this. We're going to have this issue generally, not just with the binary literals.

I think we should have some logic that says "if it starts with 0x[" or "if it starts with 0b[" this it's definitely a binary literal and if it doesn't parse, we should error there.

**Comment by jntrnr** [comment]
I'm going to land this and also file an issue for what @sholderbach mentioned about improving parser errors.

**Pull request merged**